### PR TITLE
#2109 Add Spotless Maven Plugin

### LINKED ISSUE #1209: [BUG] Bedrock chat model fails with validation exception when non-ascii character is present in messages
**Describe the bug**
Aws bedrock chat model returns validation exception when accented characters like á is present in the input messages.

```
result = {ValidationException@22166} Method threw 'software.amazon.awssdk.services.bedrockruntime.model.ValidationException' exception.
 stackTrace = {StackTraceElement[49]@22173} 
  0 = {StackTraceElement@22182} "software.amazon.awssdk.core.internal.http.CombinedResponseHandler.handleErrorResponse(CombinedResponseHandler.java:125)"
  1 = {StackTraceElement@22183} "software.amazon.awssdk.core.internal.http.CombinedResponseHandler.handleResponse(CombinedResponseHandler.java:82)"
  2 = {StackTraceElement@22184} "software.amazon.awssdk.core.internal.http.CombinedResponseHandler.handle(CombinedResponseHandler.java:60)"
  3 = {StackTraceElement@22185} "software.amazon.awssdk.core.internal.http.CombinedResponseHandler.handle(CombinedResponseHandler.java:41)"
  4 = {StackTraceElement@22186} "software.amazon.awssdk.core.internal.http.pipeline.stages.HandleResponseStage.execute(HandleResponseStage.java:50)"
  5 = {StackTraceElement@22187} "software.amazon.awssdk.core.internal.http.pipeline.stages.HandleResponseStage.execute(HandleResponseStage.java:38)"
  6 = {StackTraceElement@22188} "software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)"
  7 = {StackTraceElement@22189} "software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallAttemptTimeoutTrackingStage.execute(ApiCallAttemptTimeoutTrackingStage.java:72)"
  8 = {StackTraceElement@22190} "software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallAttemptTimeoutTrackingStage.execute(ApiCallAttemptTimeoutTrackingStage.java:42)"
  9 = {StackTraceElement@22191} "software.amazon.awssdk.core.internal.http.pipeline.stages.TimeoutExceptionHandlingStage.execute(TimeoutExceptionHandlingStage.java:78)"
  10 = {StackTraceElement@22192} "software.amazon.awssdk.core.internal.http.pipeline.stages.TimeoutExceptionHandlingStage.execute(TimeoutExceptionHandlingStage.java:40)"
  11 = {StackTraceElement@22193} "software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallAttemptMetricCollectionStage.execute(ApiCallAttemptMetricCollectionStage.java:55)"
  12 = {StackTraceElement@22194} "software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallAttemptMetricCollectionStage.execute(ApiCallAttemptMetricCollectionStage.java:39)"
  13 = {StackTraceElement@22195} "software.amazon.awssdk.core.internal.http.pipeline.stages.RetryableStage.execute(RetryableStage.java:81)"
  14 = {StackTraceElement@22196} "software.amazon.awssdk.core.internal.http.pipeline.stages.RetryableStage.execute(RetryableStage.java:36)"
  15 = {StackTraceElement@22197} "software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)"
  16 = {StackTraceElement@22198} "software.amazon.awssdk.core.internal.http.StreamManagingStage.execute(StreamManagingStage.java:56)"
  17 = {StackTraceElement@22199} "software.amazon.awssdk.core.internal.http.StreamManagingStage.execute(StreamManagingStage.java:36)"
  18 = {StackTraceElement@22200} "software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallTimeoutTrackingStage.executeWithTimer(ApiCallTimeoutTrackingStage.java:80)"
  19 = {StackTraceElement@22201} "software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallTimeoutTrackingStage.execute(ApiCallTimeoutTrackingStage.java:60)"
  20 = {StackTraceElement@22202} "software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallTimeoutTrackingStage.execute(ApiCallTimeoutTrackingStage.java:42)"
  21 = {StackTraceElement@22203} "software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallMetricCollectionStage.execute(ApiCallMetricCollectionStage.java:50)"
  22 = {StackTraceElement@22204} "software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallMetricCollectionStage.execute(ApiCallMetricCollectionStage.java:32)"
  23 = {StackTraceElement@22205} "software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)"
  24 = {StackTraceElement@22206} "software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)"
  25 = {StackTraceElement@22207} "software.amazon.awssdk.core.internal.http.pipeline.stages.ExecutionFailureExceptionReportingStage.execute(ExecutionFailureExceptionReportingStage.java:37)"
  26 = {StackTraceElement@22208} "software.amazon.awssdk.core.internal.http.pipeline.stages.ExecutionFailureExceptionReportingStage.execute(ExecutionFailureExceptionReportingStage.java:26)"
  27 = {StackTraceElement@22209} "software.amazon.awssdk.core.internal.http.AmazonSyncHttpClient$RequestExecutionBuilderImpl.execute(AmazonSyncHttpClient.java:224)"
  28 = {StackTraceElement@22210} "software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.invoke(BaseSyncClientHandler.java:103)"
  29 = {StackTraceElement@22211} "software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.doExecute(BaseSyncClientHandler.java:173)"
  30 = {StackTraceElement@22212} "software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.lambda$execute$1(BaseSyncClientHandler.java:80)"
  31 = {StackTraceElement@22213} "software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.measureApiCallSuccess(BaseSyncClientHandler.java:182)"
  32 = {StackTraceElement@22214} "software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.execute(BaseSyncClientHandler.java:74)"
  33 = {StackTraceElement@22215} "software.amazon.awssdk.core.client.handler.SdkSyncClientHandler.execute(SdkSyncClientHandler.java:45)"
  34 = {StackTraceElement@22216} "software.amazon.awssdk.awscore.client.handler.AwsSyncClientHandler.execute(AwsSyncClientHandler.java:53)"
  35 = {StackTraceElement@22217} "software.amazon.awssdk.services.bedrockruntime.DefaultBedrockRuntimeClient.invokeModel(DefaultBedrockRuntimeClient.java:149)"
  36 = {StackTraceElement@22218} "dev.langchain4j.model.bedrock.internal.AbstractBedrockChatModel.invoke(AbstractBedrockChatModel.java:141)"
  37 = {StackTraceElement@22219} "dev.langchain4j.model.bedrock.BedrockAnthropicMessageChatModel.lambda$generate$0(BedrockAnthropicMessageChatModel.java:69)"
  38 = {StackTraceElement@22220} "dev.langchain4j.internal.RetryUtils$RetryPolicy.withRetry(RetryUtils.java:192)"
  39 = {StackTraceElement@22221} "dev.langchain4j.internal.RetryUtils.withRetry(RetryUtils.java:229)"
  40 = {StackTraceElement@22222} "dev.langchain4j.model.bedrock.BedrockAnthropicMessageChatModel.generate(BedrockAnthropicMessageChatModel.java:69)"
 awsErrorDetails = {AwsErrorDetails@22169} "AwsErrorDetails(errorMessage=Malformed input request, please reformat your input and try again., errorCode=ValidationException, serviceName=BedrockRuntime)"
 clockSkew = {Duration@21122} "PT0S"
 requestId = "4c40820b-8c19-4e90-97ee-225f8642398e"
 extendedRequestId = null
 statusCode = 400
 backtrace = {Object[5]@22171} 
 detailMessage = "Malformed input request, please reformat your input and try again."
 cause = null
 depth = 49
 suppressedExceptions = {Collections$EmptyList@21127}  size = 0
```

**To Reproduce**
To reproduce this issue, use non-ascii character with AWS Bedrock chat model(like claude 3). It needs to call AbstractBedrockChatModel.java's generate function.

**Expected behavior**
It should work regardless of what character is in the string.

**Root Cause & Fix**
The root cause for the issue is that we are using the default charset for the request in AbstractBedrockChatModel.java's invoke function. In order to fix this issue, we need to change the "Charset.defaultCharset()" to "StandardCharsets.UTF_8". 

**Please complete the following information:**
- LangChain4j version: 0.29
- LLM(s) used: Bedrock Claude 3
- Java version: 11

### LINKED ISSUE #2109: [BUG] Code style autoformatter and checker is missing
**Describe the bug**
Code style / formatting enforcer is missing. This cause [unnecessary traction during code reviews](https://github.com/langchain4j/langchain4j/pull/2063#discussion_r1839840416).

**Log and Stack trace**

**To Reproduce**

**Expected behavior**

- Code formatter is enabled and automatically formatting code before commit. (`mvn spotless:apply`)
- Code formatting checkhing is enabled as PR validation step (`mvn spotless:check`)
- If PR does not meet the code style / formatting requirements, PR is red and not a subject of PR reviews
- Contributors are [not getting angry 😠 and frustrated](https://github.com/langchain4j/langchain4j/pull/2064#discussion_r1840982950) and keep their motivation. Reverting formatting changes does not seem to be an adequate use of their time in 2024.

I recommend to have a look at [spotless-maven-plugin](https://github.com/diffplug/spotless/tree/main/plugin-maven)

**Please complete the following information:**
- LangChain4j version:
- LLM(s) used: 
- Java version: 
- Spring Boot version (if applicable): 

**Additional context**

### PR DESCRIPTION
**Pull request by kpavlov**
## Issue
Adds format consistency check for changed files. Closes #2109 

## Change

- Add [Spotless Maven plugin](https://github.com/diffplug/spotless) for code java (eclipse java format) and pom.xml formatting
- Added the `<project.root.dir>` property across multiple POM files to standardize the directory structure reference. This ensures consistent project configurations and reduces redundancy in defining the root project directory. Adjusted existing XML formatting minimally to accommodate the new property integration.
- Cleanup duplicate settings in pgvector module
- Updated pom.xml files were reformatted according to [defaults](https://github.com/diffplug/spotless/tree/main/plugin-maven#sortpom) + 4 spaces
- Adding linting and formatting commands to the `Makefile`, updating the `pom.xml` files for better project configuration, and enhancing the `CONTRIBUTING.md` guidelines.

## General checklist

- [x] There are no breaking changes
- [ ] I have added unit and integration tests for my change
- [ ] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [ ] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green

- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)

## Checklist for adding new maven module

- [ ] I have added my new module in the root `pom.xml` and `langchain4j-bom/pom.xml`

## Checklist for adding new embedding store integration

- [ ] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreRemovalIT` that extends from `EmbeddingStoreWithRemovalIT`

## Checklist for changing existing embedding store integration

- [ ] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

### CODE REVIEW

File: langchain4j-parent/pom.xml [hunk 1]
```diff
@@ -436,6 +479,25 @@
...
+                <plugin>
+                    <groupId>com.diffplug.spotless</groupId>
+                    <artifactId>spotless-maven-plugin</artifactId>
+                    <version>2.43.0</version>
+                    <configuration>
+                        <!-- optional: limit format enforcement to just the files changed by this feature branch -->
+                        <ratchetFrom>origin/main</ratchetFrom>
+                        <java>
+                            <palantirJavaFormat>
+                                <version>2.50.0</version>
```
Comment (langchain4j) [inline langchain4j-parent/pom.xml:516]: AFAIU this is different from code style you've added in `.editorconfig`? Is there a way to use `.editorconfig` here as well?

**Review by ghost — COMMENTED** [review]
@kpavlov thank you!

File: langchain4j-parent/pom.xml (hunk 1 above)
Comment (kpavlov) [reply langchain4j-parent/pom.xml:516]: Not in palantirJavaFormat or googleJavaFormat
🚫 We can try [editorconfig-maven-plugin](https://github.com/ec4j/editorconfig-maven-plugin) . It was last updated more than year ago. 
✅ Or update `.editorconfig` to match the palantir style, which is the closest to the Idea Style.
❓ Or switch to the strict GOOGLE/AOSP style

Comment (kpavlov) [reply langchain4j-parent/pom.xml:516]: - editorconfig-maven-plugin has very limited set of [supported .editorconfig properties](https://github.com/ec4j/editorconfig-maven-plugin?tab=readme-ov-file#how-it-works) and not suppporting custome `ij_*` settings,
- editorconfig-maven-plugin does not support ratchetFrom, so a Big-Bang reformat is required.

Not an option.

Comment (langchain4j) [reply langchain4j-parent/pom.xml:516]: So there is no point to have `.editorconfig` if we want to automate checks and reformatting? AFAIU there should be a single "source of truth"

Comment (kpavlov) [reply langchain4j-parent/pom.xml:516]: There is a point  to have `.editorconfig`: to help IDEs with formatting.
@langchain4j , I understand that palantirJavaFormat style might not be exactly what you want to see. I encourage you to tune the formatter and checker, so they meet your code style. No one could do it better than you 🤗

File: langchain4j-pgvector/pom.xml [hunk 2]
```diff
@@ -158,7 +147,7 @@
                     </formats>
                     <java>
                         <googleJavaFormat>
```
Comment (langchain4j) [inline langchain4j-pgvector/pom.xml]: I guess we should remove this now

**Review by ghost — APPROVED** [review]
@kpavlov thanks a lot!

**Pull request merged**